### ISSUE TITLE
OperationalError: too many SQL variables

### ISSUE THREAD

**Issue by radinamatic** [learningequality/kolibri#2419]
## Summary

Happens in VM and my local dev repo, where I have imported CK-12 channel from USB drive. Channel isn't showing as `Featured in...` in `Recommended` tab with the below errors in the console, but its content can be browsed correctly in `Topics`.

![fullscreen_10_10_17_03_59](https://user-images.githubusercontent.com/1457929/31458243-cf9f4206-aebf-11e7-8c00-7ad4059b2270.png)

![fullscreen_10_10_17_03_54](https://user-images.githubusercontent.com/1457929/31458346-1aef27ee-aec0-11e7-98ac-47e6a342a3b1.png)

## System information

 - Version: latest develop
 - Operating system: Win VM and develop on OSX
 - Browser: all

## Traceback or relevant snippet from server.log or browser console

```
Environment:

Request Method: GET
Request URL: http://127.0.0.1:8000/api/contentnode/all_content/?channel_id=1d8f6d84618153c18c695d85074952a7&1507745570490=1507745570490

Django Version: 1.9.7
Python Version: 2.7.14
Installed Applications:
[u'kolibri.core',
 u'django.contrib.admin',
 u'django.contrib.auth',
 u'django.contrib.contenttypes',
 u'django.contrib.sessions',
 u'django.contrib.messages',
 u'django.contrib.staticfiles',
 u'kolibri.auth.apps.KolibriAuthConfig',
 u'kolibri.content',
 u'kolibri.logger',
 u'kolibri.tasks.apps.KolibriTasksConfig',
 u'kolibri.core.webpack',
 u'kolibri.core.exams',
 u'kolibri.core.device',
 u'kolibri.core.discovery',
 u'rest_framework',
 u'django_js_reverse',
 u'jsonfield',
 u'morango',
 u'kolibri.plugins.management',
 u'kolibri.plugins.learn',
 u'kolibri.plugins.document_pdf_render',
 u'kolibri.plugins.html5_app_renderer',
 u'kolibri.plugins.media_player',
 u'kolibri.plugins.setup_wizard',
 u'kolibri.plugins.coach',
 u'kolibri.plugins.user',
 u'kolibri.plugins.style_guide',
 u'kolibri_exercise_perseus_plugin']
Installed Middleware:
(u'django.contrib.sessions.middleware.SessionMiddleware',
 u'kolibri.core.device.middleware.KolibriLocaleMiddleware',
 u'django.middleware.common.CommonMiddleware',
 u'django.middleware.csrf.CsrfViewMiddleware',
 u'kolibri.plugins.setup_wizard.middleware.SetupWizardMiddleware',
 u'kolibri.auth.middleware.CustomAuthenticationMiddleware',
 u'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
 u'django.contrib.messages.middleware.MessageMiddleware',
 u'django.middleware.clickjacking.XFrameOptionsMiddleware',
 u'django.middleware.security.SecurityMiddleware')



Traceback:

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  149.                     response = self.process_exception_by_middleware(e, request)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  147.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/views/decorators/csrf.py" in wrapped_view
  58.         return view_func(*args, **kwargs)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/rest_framework/viewsets.py" in view
  87.             return self.dispatch(request, *args, **kwargs)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  466.             response = self.handle_exception(exc)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/rest_framework/views.py" in dispatch
  463.             response = handler(request, *args, **kwargs)

File "kolibri/content/api.py" in all_content
  280.         return Response(serializer.data)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/rest_framework/serializers.py" in data
  674.         ret = super(ListSerializer, self).data

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/rest_framework/serializers.py" in data
  239.                 self._data = self.to_representation(self.instance)

File "kolibri/content/serializers.py" in to_representation
  170.         if not data:

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in __nonzero__
  266.         return type(self).__bool__(self)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in __bool__
  262.         self._fetch_all()

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in _fetch_all
  1076.             self._prefetch_related_objects()

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in _prefetch_related_objects
  656.         prefetch_related_objects(self._result_cache, self._prefetch_related_lookups)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in prefetch_related_objects
  1457.                 obj_list, additional_lookups = prefetch_one_level(obj_list, prefetcher, lookup, level)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in prefetch_one_level
  1556.         prefetcher.get_prefetch_queryset(instances, lookup.get_current_queryset(level)))

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py" in get_prefetch_queryset
  543.             for rel_obj in queryset:

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in __iter__
  258.         self._fetch_all()

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in _fetch_all
  1074.             self._result_cache = list(self.iterator())

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/query.py" in __iter__
  52.         results = compiler.execute_sql()

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  848.             cursor.execute(sql, params)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  79.             return super(CursorDebugWrapper, self).execute(sql, params)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  64.                 return self.cursor.execute(sql, params)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/utils.py" in __exit__
  95.                 six.reraise(dj_exc_type, dj_exc_value, traceback)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  64.                 return self.cursor.execute(sql, params)

File "/Users/mac/.venvs/kolibri-sans2/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py" in execute
  323.         return Database.Cursor.execute(self, query, params)

Exception Type: OperationalError at /api/contentnode/all_content/
Exception Value: too many SQL variables
```

## How to reproduce

1. [Download ZIP](https://drive.google.com/file/d/0B3ZvXdQEq36jM0lJSm96OUNmMjA/view?usp=sharing) (beware, 6GB)
1. Unzip and do a local import 
1. Go to `Learn`

**Comment by radinamatic** [comment]
@lyw07 Did you try importing this channel (this is on the link from Slack)? Could you reproduce this error?

**Comment by rtibbles** [comment]
@radinamatic was this on latest develop, or on @lyw07's PR? Just want to verify because they may not be completely identical.

If the latter, please double check and replicate on latest develop also.

**Comment by radinamatic** [comment]
@rtibbles I was extant on yesterday's develop, I'll pull again and restart.

**Comment by rtibbles** [comment]
No need to retest then.

**Comment by rtibbles** [comment]
This is only on the CK12 channel though, yes? Smaller channels, not a problem?

**Comment by radinamatic** [comment]
I didn't see it on other channels, only after importing this one. I'll rename the `/content/` folder and re-test just with smaller ones.

**Comment by radinamatic** [comment]
Here is the sequence (on the just-updated-develop) from login to the error, I recorded it because there was another error in console during login, and I'm not sure if it can be related...

![sql](https://user-images.githubusercontent.com/1457929/31460588-3c7da816-aec7-11e7-802b-596435d93ee6.gif)

**Comment by rtibbles** [comment]
The sign in error is unrelated (and not a bug, just always happens when you log in as a non learner to simplified login).

**Comment by radinamatic** [comment]
Restarted with the fresh `./kolibri/` folder and imported my small testing channel `8b4d3e6d3d4842ba8ea658335b5dd252`, some `attempt log` red in the console, but no mention of the SQL:

![fullscreen_11_10_17_21_33](https://user-images.githubusercontent.com/1457929/31462492-3e6d0bca-aecd-11e7-958e-9f9298c6dbfe.png)

Then I tried importing **Portion control** `fb51dae6df7545af8455aa3a0c32048d` and another small one `95a52b386f2c485cb97dd60901674a98` and I'm getting again the unspecified `Transfer failed. Please try again.` in UI... :rage: 

![fullscreen_11_10_17_21_47](https://user-images.githubusercontent.com/1457929/31462718-e1df48a4-aecd-11e7-9b87-15084bd26472.png)

Switching to testing other stuff in VMs for a while so I can cool off... :unamused:

**Comment by rtibbles** [comment]
@radinamatic can you send the `server.log` file so we can see what is causing the failed transfer there?

**Comment by radinamatic** [comment]
[kolibri.log](https://github.com/learningequality/kolibri/files/1377041/kolibri.log)

[debug.log](https://github.com/learningequality/kolibri/files/1377043/debug.log)

I don't see `server.log`... :thinking:

**Comment by radinamatic** [comment]
@ivanistheone Do these look like unhelpful logs you were seeing with the same issue we had on demo servers last week?

**Comment by ivanistheone** [comment]
@radinamatic Yes, the above errors resemble what we were seeing the other day when re-importing a channel.

I ran into the bug this morning again (still using 0.6.0beta4). I was able to import the channel initially w/o problems, but then the **second time importing got the error** message in the frontend, and this was in the logs:

```
[37mINFO     Downloading data for channel id 310ec19477d15cf7b9fed98551ba1e1f^[[0m
INFO:kolibri.content.management.commands.importchannel:Downloading data for channel id 310ec19477d15cf7b9fed98551ba1e1f
^M  0%|          | 0/1515520 [00:00<?, ?it/s]^M100%|##########| 1515520/1515520 [00:00<00:00, 6595671.06it/s]^[[37mINFO     Channel 310ec19477d15cf7b9fed98551ba1e1f already exists in database, cleaning up ContentNodes^[[0m
INFO:kolibri.content.utils.channel_import:Channel 310ec19477d15cf7b9fed98551ba1e1f already exists in database, cleaning up ContentNodes
^[[37mINFO     Importing ContentTag data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ContentTag data
^[[37mINFO     Importing ContentNode_has_prerequisite data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ContentNode_has_prerequisite data
^[[37mINFO     Importing ContentNode_related data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ContentNode_related data
^[[37mINFO     Importing ContentNode_tags data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ContentNode_tags data
^[[37mINFO     Importing ContentNode data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ContentNode data
^[[37mINFO     Importing Language data^[[0m
INFO:kolibri.content.utils.channel_import:Importing Language data
^[[37mINFO     Importing File data^[[0m
INFO:kolibri.content.utils.channel_import:Importing File data
^[[37mINFO     Importing LocalFile data^[[0m
INFO:kolibri.content.utils.channel_import:Importing LocalFile data
^[[37mINFO     Importing License data^[[0m
INFO:kolibri.content.utils.channel_import:Importing License data
^[[37mINFO     Importing AssessmentMetaData data^[[0m
INFO:kolibri.content.utils.channel_import:Importing AssessmentMetaData data
^[[37mINFO     Importing ChannelMetadata data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ChannelMetadata data
^[[37mINFO     Importing ContentNode_Deferred_author_available_channel_id7ca2f1a6a5385fe41b9aa2439213d8b9 data^[[0m
INFO:kolibri.content.utils.channel_import:Importing ContentNode_Deferred_author_available_channel_id7ca2f1a6a5385fe41b9aa2439213d8b9 data
```

After restarting Kolibri, the channel appeared (as if operation completed w/o error), but this was in the logs:

```
WARNING:barbequeue.common.utils.MESSAGEPROCESSOR[yankee-football-indigo-mobile]:Got an exception running <bound method WorkerBackend.process_messages of <barbequeue.worker.backends.inmem.WorkerBackend object at 0x7eff506a6f10>>: u'8e76ba9c9163418da9790978da4faaf7'
^[[37mINFO     Downloading data for channel id 310ec19477d15cf7b9fed98551ba1e1f^[[0m
INFO:kolibri.content.management.commands.importchannel:Downloading data for channel id 310ec19477d15cf7b9fed98551ba1e1f
^M  0%|          | 0/1515520 [00:00<?, ?it/s]^M100%|##########| 1515520/1515520 [00:00<00:00, 6303245.05it/s]^[[37mINFO     Channel 310ec19477d15cf7b9fed98551ba1e1f already exists in database, cleaning up ContentNodes^[[0m
INFO:kolibri.content.utils.channel_import:Channel 310ec19477d15cf7b9fed98551ba1e1f already exists in database, cleaning up ContentNodes
```

Happy to update le-060-beta if someone sends a link to post-fixes pex my way.

**Comment by radinamatic** [comment]
That seems just what is happening on my develop... I restarted it after the failed imports, and as you say, the 2 channels appear in `Device/Content`, but not on `Learn`:

![fullscreen_12_10_17_00_40](https://user-images.githubusercontent.com/1457929/31470934-84c85f3e-aee6-11e7-9cc5-9f3ebc28f603.png)

```
Last login: Tue Oct 10 17:31:11 on ttys001
Dmac:~ mac$ cd LE/kolibri
Dmac:kolibri mac$ source ~/.venvs/kolibri-sans2/bin/activate
(kolibri-sans2) Dmac:kolibri mac$ kolibri --debug manage devserver --webpack
WARNING:root:No C Extensions available for this platform.

DEBUG    Debug mode is on!
DEBUG:kolibri.utils.cli:Debug mode is on!
INFO     Starting webpack process from Django runserver command
INFO:kolibri.core.webpack.management.commands.devserver:Starting webpack process from Django runserver command
INFO     Django Runserver command has spawned a Webpack watcher process on pid 16108
INFO:kolibri.core.webpack.management.commands.devserver:Django Runserver command has spawned a Webpack watcher process on pid 16108
INFO     Importing ContentTag data
INFO:kolibri.content.utils.channel_import:Importing ContentTag data
INFO     Importing ContentNode_has_prerequisite data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_has_prerequisite data
INFO     Importing ContentNode_related data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_related data
INFO     Importing ContentNode_tags data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_tags data
INFO     Importing ContentNode data
INFO:kolibri.content.utils.channel_import:Importing ContentNode data
INFO     Importing Language data
INFO:kolibri.content.utils.channel_import:Importing Language data
INFO     Importing File data
INFO:kolibri.content.utils.channel_import:Importing File data
INFO     Importing LocalFile data
INFO:kolibri.content.utils.channel_import:Importing LocalFile data
INFO     Importing AssessmentMetaData data
INFO:kolibri.content.utils.channel_import:Importing AssessmentMetaData data
INFO     Importing ChannelMetadata data
INFO:kolibri.content.utils.channel_import:Importing ChannelMetadata data
INFO     Setting availability of File objects based on LocalFile availability
INFO:kolibri.content.utils.annotation:Setting availability of File objects based on LocalFile availability
INFO     Setting availability of non-topic ContentNode objects based on File availability
INFO:kolibri.content.utils.annotation:Setting availability of non-topic ContentNode objects based on File availability
INFO     Importing ContentTag data
INFO:kolibri.content.utils.channel_import:Importing ContentTag data
INFO     Importing ContentNode_has_prerequisite data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_has_prerequisite data
INFO     Importing ContentNode_related data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_related data
INFO     Importing ContentNode_tags data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_tags data
INFO     Importing ContentNode data
INFO:kolibri.content.utils.channel_import:Importing ContentNode data
INFO     Importing Language data
INFO:kolibri.content.utils.channel_import:Importing Language data
INFO     Importing File data
INFO:kolibri.content.utils.channel_import:Importing File data
INFO     Importing LocalFile data
INFO:kolibri.content.utils.channel_import:Importing LocalFile data
INFO     Importing AssessmentMetaData data
INFO:kolibri.content.utils.channel_import:Importing AssessmentMetaData data
INFO     Importing ChannelMetadata data
INFO:kolibri.content.utils.channel_import:Importing ChannelMetadata data
INFO     Setting availability of File objects based on LocalFile availability
INFO:kolibri.content.utils.annotation:Setting availability of File objects based on LocalFile availability
INFO     Setting availability of non-topic ContentNode objects based on File availability
INFO:kolibri.content.utils.annotation:Setting availability of non-topic ContentNode objects based on File availability
INFO     Setting availability of LocalFile objects based on disk availability
INFO:kolibri.content.utils.annotation:Setting availability of LocalFile objects based on disk availability
yarn run v1.1.0
INFO     Setting availability of 108 LocalFile objects based on passed in checksums
INFO:kolibri.content.utils.annotation:Setting availability of 108 LocalFile objects based on passed in checksums
INFO     Setting availability of File objects based on LocalFile availability
INFO:kolibri.content.utils.annotation:Setting availability of File objects based on LocalFile availability
INFO     Setting availability of non-topic ContentNode objects based on File availability
INFO:kolibri.content.utils.annotation:Setting availability of non-topic ContentNode objects based on File availability
INFO     Setting availability of ContentNode objects with children for 5 levels
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for 5 levels
INFO     Setting availability of ContentNode objects with children for level 5
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 5
INFO     Setting availability of ContentNode objects with children for level 4
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 4
INFO     Setting availability of ContentNode objects with children for level 3
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 3
INFO     Setting availability of ContentNode objects with children for level 2
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 2
INFO     Setting availability of ContentNode objects with children for level 1
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 1
$ node ./frontend_build/src/webpackdevserver.js
WARNING:root:No C Extensions available for this platform.

DEBUG    Debug mode is on!
DEBUG:kolibri.utils.cli:Debug mode is on!
Performing system checks...

System check identified no issues (1 silenced).
October 12, 2017 - 00:37:15
Django version 1.9.7, using settings 'kolibri.deployment.default.settings.base'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
WARNING:root:No C Extensions available for this platform.

WARNING  assets/src/module.js not found for plugin exercise_perseus_render_module.
WARNING:kolibri.core.webpack.hooks:assets/src/module.js not found for plugin exercise_perseus_render_module.
INFO     Writing webpack_json output to /var/folders/9g/_0n3kc0j15z4jt4bnfxxf_th0000gn/T/117912-16114-1ayj7m1.42c4ayk3xr.json
INFO:kolibri.core.webpack.management.commands.webpack_json:Writing webpack_json output to /var/folders/9g/_0n3kc0j15z4jt4bnfxxf_th0000gn/T/117912-16114-1ayj7m1.42c4ayk3xr.json
Hash: 70facacbd59ecf07a37f38c95fac7cbe9bcc4b6d19e7869f93b7cbdde7cf94462b011db1e42a34bff0962ceae5c3e9004d4d129340b16caf92fa3b23362dd5b002d9cc562a7dd94cb94206c277c4a7106eeef685b56738001b8f821df31211fdb2934cef2d4d67d4b25a057962f4
Version: webpack 3.6.0
Child default_frontend:
    Hash: 70facacbd59ecf07a37f
    Time: 63978ms
...
```

**Comment by rtibbles** [comment]
Sounds like something broke during file download? If the channel gets imported, but the download process errors out, then the channel won't get annotated.

It will still show in the content management, but won't show to learners, as there is no content available.

**Comment by rtibbles** [comment]
Fixed in #2424
